scipy-stubs/interpolate/_bsplines.pyi

```diff
@@ -11,66 +11,109 @@ __all__ = ["BSpline", "make_interp_spline", "make_lsq_spline", "make_smoothing_s
-    def tck(self, /) -> tuple[onp.Array1D[np.float64], onp.Array[onp.AtLeast1D, _SCT_co], int]: ...
+    def tck(self, /) -> tuple[onp.Array1D[np.float64], onp.Array[onp.AtLeast1D, _CT_co], int]: ...
 
     #
+    @overload
     def __init__(
-        self,
+        self: BSpline[np.float64],
         /,
         t: onp.ToFloat1D,
         c: onp.ToFloatND,
         k: op.CanIndex,
         extrapolate: _Extrapolate = True,
         axis: op.CanIndex = 0,
     ) -> None: ...
+    @overload
+    def __init__(
+        self: BSpline[np.complex128],
+        /,
+        t: onp.ToFloat1D,
+        c: onp.ToJustComplexND,
+        k: op.CanIndex,
+        extrapolate: _Extrapolate = True,
+        axis: op.CanIndex = 0,
+    ) -> None: ...
+    @overload
+    def __init__(
+        self: BSpline[Any],
+        /,
+        t: onp.ToFloat1D,
+        c: onp.ToComplex1D,
+        k: op.CanIndex,
+        extrapolate: _Extrapolate = True,
+        axis: op.CanIndex = 0,
+    ) -> None: ...
 
-    #
+    # NOTE: Complex `x` will unsafely be cast to `float64`, even if the coefficients are complex
     def __call__(
         self,
         /,
-        x: onp.ToFloat | onp.ToFloatND,
+        x: onp.ToComplex | onp.ToComplexND,
         nu: int = 0,
         extrapolate: _Extrapolate | None = None,
-    ) -> onp.ArrayND[_SCT_co]: ...
+    ) -> onp.ArrayND[_CT_co]: ...
 
     #
     def derivative(self, /, nu: int = 1) -> Self: ...
     def antiderivative(self, /, nu: int = 1) -> Self: ...
+
+    #
     def insert_knot(self, /, x: onp.ToFloat, m: op.CanIndex = 1) -> Self: ...
-    def integrate(self, /, a: onp.ToFloat, b: onp.ToFloat, extrapolate: _Extrapolate | None = None) -> onp.ArrayND[_SCT_co]: ...
+
+    # NOTE: `integrate` will raise a (cryptic) `ValueError` for complex coefficients
+    def integrate(
+        self: BSpline[np.float64],
+        /,
+        a: onp.ToFloat,
+        b: onp.ToFloat,
+        extrapolate: _Extrapolate | None = None,
+    ) -> onp.ArrayND[np.float64]: ...
 
     #
+    @overload
+    @classmethod
+    def basis_element(cls, t: onp.ToFloatND, extrapolate: _Extrapolate = True) -> BSpline[np.float64]: ...
+    @overload
     @classmethod
-    def basis_element(cls, t: onp.ToFloat1D, extrapolate: _Extrapolate = True) -> Self: ...
+    def basis_element(cls, t: onp.ToJustComplexND, extrapolate: _Extrapolate = True) -> BSpline[np.complex128]: ...
+    @overload
     @classmethod
-    def from_power_basis(cls, pp: CubicSpline, bc_type: _BCType = "not-a-knot") -> Self: ...
+    def basis_element(cls, t: onp.ToComplexND, extrapolate: _Extrapolate = True) -> BSpline[Any]: ...
+
+    #
+    @classmethod
+    def from_power_basis(cls, pp: CubicSpline[_CT_co], bc_type: _BCType = "not-a-knot") -> Self: ...
+
+    #
     @classmethod
     def construct_fast(
         cls,
-        t: onp.Array1D[np.float64],
-        c: onp.Array[onp.AtLeast1D, _SCT_co],
+        t: onp.ArrayND[np.float64],
+        c: onp.ArrayND[_CT_co],
```

Comment:
`construct_fast` gets exactly the same arguments as `__init__`, should probably be typed the same?

Reply:
It skips the conversion and validation steps, so it doesn't convert array-likes to arrays, for example. That's why `t` and `c` are annotated to have (almost) the same type as their corresponding attributes.